cylc/uiserver/schema.py

```diff
@@ -396,78 +415,299 @@ def run_task_query(conn, workflow):
-            'mean_total_time': row[24],
-            'max_total_time': row[25],
-            'std_dev_total_time': (row[26] - row[24] ** 2) ** 0.5,
+            'min_total_time': row[24],
+            'mean_total_time': row[25],
+            'max_total_time': row[26],
+            'std_dev_total_time': (row[27] - row[25] ** 2) ** 0.5,
             # Prevents null entries when there are too few tasks for quartiles
-            'total_quartiles': [row[27],
-                                row[27] if row[28] is None else row[28],
-                                row[27] if row[29] is None else row[29]],
+            'total_quartiles': [row[28],
+                                row[28] if row[29] is None else row[29],
+                                row[28] if row[30] is None else row[30]],
 
-            'count': row[30]
+            'count': row[31]
         })
 
     return tasks
 
 
-def run_jobs_query(conn, workflow, tasks):
+_JOB_STATUS_TO_STATE = {
+    # task_status: (submit_status, run_status, time_run)
+    TASK_STATUS_SUBMITTED: (0, None, None),
+    TASK_STATUS_SUBMIT_FAILED: (1, None, None),
+    TASK_STATUS_RUNNING: (0, None, True),
+    TASK_STATUS_SUCCEEDED: (0, 0, True),
+    TASK_STATUS_FAILED: (0, 1, True),
+}
+
+
+def _status_to_state(
+    status: str
+) -> Tuple[Optional[int], Optional[int], Optional[bool]]:
+    """Derive job state attributes from job status.
+
+    The time_run cannot be derived from the status so is returned as a boolean.
+
+    Args:
+        status: The job status, e.g. "submitted" or "running".
+
+    Returns:
+        (submit_status, run_status, time_run)
+
+        submit_status:
+            * 0 = successful submission
+            * int = failed submission
+            * None = no submission attempt
+        run_status:
+            * 0 = successful execution
+            * int = failed execution
+            * None = no execution attempt
+        time_run:
+            * True if the job would have started running
+
+    Examples:
+        >>> _status_to_state('running')
+        (0, None, True)
+        >>> _status_to_state('waiting')
+        (None, None, None)
+
+        >>> _state_to_status(*_status_to_state('succeeded'))
+        'succeeded'
+
+    """
+    return _JOB_STATUS_TO_STATE.get(status, (None, None, None))
+
+
+def _state_to_status(
+    submit_status: Optional[int],
+    run_status: Optional[int],
+    time_run: Optional[str],
+) -> str:
+    """Derive job status from state attributes.
+
+    Args:
+        submit_status: Exit code from job submission.
+        run_status: Exit code from job execution.
+        time_run: The time the job reported that it started running.
+            (note this is interpreted as a bool, the exact value is ignored).
+
+    Returns:
+        status: The job status, e.g "submitted" or "running".
+
+    Examples:
+        >>> _state_to_status(0, None, True)
+        'running'
+        >>> _state_to_status(None, None, None)
+        'waiting'
```

Comment:
IMO You should test all possibilites: This logic gives me a headache (Submit status only makes sense when it's an exit code, but 1 is truthy in python). I think that full tests are good for documentation:

```suggestion
        >>> _state_to_status(None, None, None)
        'waiting'
        >>> _state_to_status(1, None, None)
        'submit-failed'
        >>> _state_to_status(0, None, True)
        'running'
        >>> _state_to_status(0, 0, True)
        'succeeded'
        >>> _state_to_status(0, 1, True)
        'failed'
```

Reply:
> IMO You should test all possibilites

If I did that I would just be replicating the dictionary above, essentially:

```
x  = {
   'a': 1,
    'b': 2,
   'c': 3
}

"""
>>> x['a']
1
>>> x['b']
2
>>> x['c']
3
"""
```

Which would only really be testing that the dictionary matches the test, not that the code is actually doing the right thing.